File: mobile/lib/core/analytics/adapters/posthog_adapter.dart
import 'package:posthog_flutter/posthog_flutter.dart';
import '../analytics_interface.dart';

/// PostHog analytics adapter.
///
/// Adapts the posthog_flutter SDK to the [AnalyticsAdapter] interface.
///
/// To add a new provider (e.g. Mixpanel):
///   1. Create adapters/mixpanel_adapter.dart implementing AnalyticsAdapter.
///   2. Register it in analytics_provider.dart factory logic.
class PostHogAnalyticsAdapter implements AnalyticsAdapter {
  PostHogAnalyticsAdapter._();

  static Future<PostHogAnalyticsAdapter> init({
    required String apiKey,
    required String host,
  }) async {
    final config = PostHogConfig(apiKey)
      ..host = host
      // Disable automatic lifecycle events — we capture events explicitly
      ..captureApplicationLifecycleEvents = false;

    await Posthog().setup(config);
    return PostHogAnalyticsAdapter._();
  }

  @override
  Future<void> capture(String event, [Map<String, Object>? properties]) async {
    await Posthog().capture(
      eventName: event,
      properties: properties,
    );
  }

  @override
  Future<void> identify(String userId, [Map<String, Object>? properties]) async {
    await Posthog().identify(
      userId: userId,
      userProperties: properties,
    );
  }

  @override
  Future<void> reset() async {
    await Posthog().reset();
  }

  @override
  Future<void> screen(String screenName, [Map<String, Object>? properties]) async {
    await Posthog().screen(
      screenName: screenName,
      properties: properties,
    );
  }

  @override
  Future<void> group(
    String groupType,
    String groupKey, [
    Map<String, Object>? properties,
  ]) async {
    await Posthog().group(
      groupType: groupType,
      groupKey: groupKey,
      groupProperties: properties,
    );
  }

  @override
  Future<bool> isFeatureFlagEnabled(
    String flagKey, {
    bool defaultValue = false,
  }) async {
    try {
      final result = await Posthog().isFeatureEnabled(flagKey);
      return result;
    } catch (_) {
      return defaultValue;
    }
  }

  @override
  Future<Map<String, Object>> getAllFeatureFlags() async {
    // posthog_flutter v4 does not expose a bulk getAllFlags method.
    // Return an empty map; override this in a custom adapter if needed.
    return {};
  }

  @override
  Future<void> flush() async {
    await Posthog().flush();
  }

  @override
  Future<void> shutdown() async {
    // posthog_flutter does not expose an explicit shutdown; flush is sufficient
    await Posthog().flush();
  }
}

File: backend/src/apps/analytics/adapters/posthog_adapter.py
"""PostHog analytics adapter.

Adapts the `posthog` Python SDK to the *AnalyticsProvider* strategy interface.
The SDK uses a background thread to batch and send events, so `capture`,
`identify`, etc. return almost instantly.  Only feature-flag lookups require
a real HTTP round-trip — those are wrapped in `asyncio.to_thread` to avoid
blocking the event loop.

To add a new provider (e.g. Mixpanel):
    1. Create ``adapters/mixpanel_adapter.py`` implementing AnalyticsProvider.
    2. Register it in ``factory.py``.
"""
import asyncio
import logging
from typing import Any

import posthog as _posthog_sdk

from src.apps.analytics.interface import AnalyticsProvider

logger = logging.getLogger(__name__)


class PostHogAdapter(AnalyticsProvider):
    """Adapter: wraps the PostHog Python SDK as an *AnalyticsProvider*."""

    def __init__(self, api_key: str, host: str) -> None:
        # The posthog module exposes a module-level client by default.
        # We configure it once here; all subsequent calls use the same client.
        _posthog_sdk.api_key = api_key
        _posthog_sdk.host = host
        # Suppress PostHog's noisy internal logs unless we're debugging
        _posthog_sdk.disabled = False
        _posthog_sdk.log = logging.getLogger("posthog")
        self._client = _posthog_sdk
        logger.info("PostHog analytics adapter initialised (host=%s)", host)

    # ------------------------------------------------------------------
    # Sending operations (fire-and-forget via SDK's background thread)
    # ------------------------------------------------------------------

    async def capture(
        self,
        distinct_id: str,
        event: str,
        properties: dict[str, Any] | None = None,
    ) -> None:
        self._client.capture(distinct_id, event, properties or {})

    async def identify(
        self,
        distinct_id: str,
        properties: dict[str, Any] | None = None,
    ) -> None:
        self._client.identify(distinct_id, properties or {})

    async def alias(self, distinct_id: str, alias: str) -> None:
        self._client.alias(distinct_id, alias)

    async def group(
        self,
        distinct_id: str,
        group_type: str,
        group_key: str,
        properties: dict[str, Any] | None = None,
    ) -> None:
        self._client.group_identify(group_type, group_key, properties or {})
        # Also associate the user with the group on their next event
        self._client.capture(
            distinct_id,
            "$groupidentify",
            {
                "$group_type": group_type,
                "$group_key": group_key,
                "$group_set": properties or {},
            },
        )

    async def page(
        self,
        distinct_id: str,
        path: str,
        properties: dict[str, Any] | None = None,
    ) -> None:
        props = {"$current_url": path, **(properties or {})}
        self._client.capture(distinct_id, "$pageview", props)

    # ------------------------------------------------------------------
    # Retrieving operations (real HTTP — run in thread to stay async)
    # ------------------------------------------------------------------

    async def get_feature_flag(
        self,
        distinct_id: str,
        flag_key: str,
        default: bool = False,
    ) -> bool | str:
        try:
            result = await asyncio.to_thread(
                self._client.feature_enabled, flag_key, distinct_id
            )
            return result if result is not None else default
        except Exception as exc:
            logger.warning("PostHog get_feature_flag error: %s", exc)
            return default

    async def get_all_feature_flags(
        self,
        distinct_id: str,
    ) -> dict[str, bool | str]:
        try:
            result = await asyncio.to_thread(
                self._client.get_all_flags, distinct_id
            )
            return result or {}
        except Exception as exc:
            logger.warning("PostHog get_all_feature_flags error: %s", exc)
            return {}

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def flush(self) -> None:
        await asyncio.to_thread(self._client.flush)

    async def shutdown(self) -> None:
        await asyncio.to_thread(self._client.shutdown)
        logger.info("PostHog analytics adapter shut down")
